[Merged by Bors] - feat(data/fintype/card_embedding): the birthday problem

ericrbg opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

- [x] depends on: #7527

New to-do list (old in edit history):

- [x] Name and document everything.
- [x] Move lemmata out of `card_embedding` and into their proper places

ericrbg committed:
🦀 all my commits are gone 🦀

my oleans got really messed up so just had to copy&pasta the file in again

ericrbg committed:
half-assed into cons, quicksave before full change

ericrbg committed:
this will work, I'm 99% sure of it

ericrbg committed:
nearly done!!!

ericrbg committed:
NO SORRYS! it works!

ericrbg committed:
woop

ericrbg committed:
final version!

ericrbg committed:
remove old extraneous lemma

ericrbg committed:
docs + we are not in 2020 anymore lol

eric-wieser commented on archive/100-theorems-list/93_birthday_problem.lean:
```diff
@@ -0,0 +1,175 @@
+local notation `‖` x `‖` := fintype.card x
+
+/-- desc_fac n k = (n + k)! / n!, but implemented in a recursive way for calculation. -/
+def desc_fac (n : ℕ) : ℕ → ℕ
```
I thought I saw @semorrison add this as a definition somewhere else, but might be misremembering.

ericrbg replied:
If you're thinking of [`pochhammer`](https://leanprover-community.github.io/mathlib_docs/ring_theory/polynomial/pochhammer.html#pochhammer), it's stated in terms of polys and needs `eval` and stuff

ericrbg replied:
Could still switch over if that's better though - I worry it won't be fast enough for `norm_num`

ericrbg committed:
fix module docstring

eric-wieser commented on archive/100-theorems-list/93_birthday_problem.lean:
```diff
@@ -0,0 +1,175 @@
+    λ f : fin n → β, λ b : β, fin.cons b f,
+
+  let equiv_classes : (fin n ↪ β) → finset (fin n.succ ↪ β) :=
+    λ f : fin n ↪ β, univ.filter (λ g : fin n.succ ↪ β, ∃ k : β, extend f k = g),
```
```suggestion
    λ f : fin n ↪ β, univ.filter (λ g : fin n.succ ↪ β, ∃ k : β, fin.cons k f = g),
```
I think you can omit `extend` entirely

ericrbg replied:
I tried doing it before, and ran into type-checker issues; the "let" was basically to force the expected type. Tried it out again in Gitpod and it seemed to be the same

Ruben-VandeVelde commented on archive/100-theorems-list/93_birthday_problem.lean:
```diff
@@ -0,0 +1,175 @@
+  induction n with n hn,
+    rw [desc_fac_zero], nontriviality (fin 0 ↪ β),
+    obtain ⟨f, g, ne⟩ := exists_pair_ne (fin 0 ↪ β),
+    exfalso, apply ne, ext x, exact x.elim0,
```
```suggestion
  { rw [desc_fac_zero], nontriviality (fin 0 ↪ β),
    obtain ⟨f, g, ne⟩ := exists_pair_ne (fin 0 ↪ β),
    exfalso, apply ne, ext x, exact x.elim0 },
```

Ruben-VandeVelde commented on archive/100-theorems-list/93_birthday_problem.lean:
```diff
@@ -0,0 +1,175 @@
+      refine ⟨⟨fin.tail f, λ _ _ h, fin.succ_inj.mp $ f.injective h⟩, _⟩,
+      suffices : ∃ (a : β), extend (fin.tail ⇑f) a = ⇑f, by simpa,
+      use f 0, simp [extend] },
+    { exact subset_univ _ },
```
```suggestion
  { apply subset.antisymm,
    { rintros f -, rw mem_bUnion,
      refine ⟨⟨fin.tail f, λ _ _ h, fin.succ_inj.mp $ f.injective h⟩, _⟩,
      suffices : ∃ (a : β), extend (fin.tail ⇑f) a = ⇑f, by simpa,
      use f 0, simp [extend] },
    { exact subset_univ _ } },
```

Ruben-VandeVelde commented on archive/100-theorems-list/93_birthday_problem.lean:
```diff
@@ -0,0 +1,175 @@
+  unfold desc_fac,
+
+  suffices : ‖fin n ↪ β‖ * (‖β‖ - n) = (‖β‖ - n.succ + n + 1) * desc_fac (‖β‖ - n.succ) n,
+    by simpa [equiv_class_size, card_univ],
```
```suggestion
  { simpa [equiv_class_size, card_univ] },
```

Ruben-VandeVelde reviewed (commented):
Some minor formatting suggestions below; I may have missed some similar cases.

ericrbg committed:
fixup style & make 100.yaml work

Co-authored-by: Ruben Van de Velde <65514131+Ruben-VandeVelde@users.noreply.github.com>

ericrbg committed:
urgh

ericrbg committed:
add general statement for α ↪ β

ericrbg committed:
remove superfluous `fintype` instances & docs

ericrbg committed:
added a couple simple corollaries

ericrbg committed:
one more TODO: relocate instances

ericrbg committed:
docstring

ericrbg committed:
moved all to correct places

ericrbg committed:
Update card_inj.lean

ericrbg committed:
fixup mistakes

eric-wieser commented on archive/100-theorems-list/93_birthday_problem.lean:
```diff
@@ -0,0 +1,21 @@
+
+theorem birthday : 2 * ‖fin 23 ↪ fin 365‖ < ‖fin 23 → fin 365‖ := by norm_num [nat.desc_fac]
+
+lemma birthday' : 2 * ‖fin 22 ↪ fin 365‖ > ‖fin 22 → fin 365‖ := by norm_num [nat.desc_fac]
```
I think for the sake of `100.yaml` which shows the theorem statement, this should be
```suggestion
theorem birthday :
  2 * ‖fin 23 ↪ fin 365‖ < ‖fin 23 → fin 365‖ ∧ 2 * ‖fin 22 ↪ fin 365‖ > ‖fin 22 → fin 365‖ :=
by split; norm_num [nat.desc_fac]
```
assuming "the birthday problem" states that "23 is the smallest number that satisfies this" as opposed to "23 satisfies this"

eric-wieser replied:
But maybe what you have is fine, looking at the statement in other provers.

ericrbg replied:
I kinda like this, I'll put this in!

eric-wieser commented on docs/100.yaml:
```diff
@@ -343,6 +343,8 @@
   title  : Pick’s Theorem
 93:
   title  : The Birthday Problem
+  decl   : fintype.card_inj
```
```suggestion
  decl   : birthday
```

ericrbg replied:
this wouldn't lint; all others that are in `archive` put it under `links`. It also kinda obscures the proof, as it's non-obvious that `norm_num` calls it; but not sure

ericrbg replied:
(yes there is a linter for this, as I found out from many CI failures...)

eric-wieser replied:
I think you should put it under links then; possibly in addition to what you have right now.

eric-wieser commented on archive/100-theorems-list/93_birthday_problem.lean:
```diff
@@ -0,0 +1,21 @@
+This file proves Theorem 93 from the [100 Theorems List](https://www.cs.ru.nl/~freek/100/).
+
+As of writing, probability is not implemented in Lean, so we instead state the birthday problem
+in terms of injective functions. We instead prove a general result about ‖α ↪ β‖.
```
```suggestion
in terms of injective functions. The general result about `fintype.card (α ↪ β)` which this proof
uses is `fintype.card_inj`.
```

eric-wieser commented on src/data/fintype/card_inj.lean:
```diff
@@ -0,0 +1,165 @@
+
+/- Establishes the cardinality of the type of all injections, if any exist.  -/
+@[simp] theorem card_inj {α β} [fintype α] [fintype β] [decidable_eq α] (h : ‖α‖ ≤ ‖β‖)
+  : ‖α ↪ β‖ = (nat.desc_fac (‖β‖ - ‖α‖) ‖α‖) :=
```
This lemma should be called `card_embedding`, since `embedding` is what `↪` is notation for, and we already have `card_equiv` to set theprecedent for that naming.

eric-wieser commented on src/data/fintype/card_inj.lean:
```diff
@@ -0,0 +1,165 @@
+end
+
+theorem card_inj'' {α β} [fintype α] [fintype β] :
+  ‖α ↪ β‖ = if ‖α‖ ≤ ‖β‖ then nat.desc_fac (‖β‖ - ‖α‖) ‖α‖ else 0 :=
```
I'd perhaps call these `card_embedding_eq_zero` and `card_embedding_eq_iff`

eric-wieser commented on src/data/fintype/card_inj.lean:
```diff
@@ -0,0 +1,165 @@
+end fintype
+
+-- just realised; is it worth registering `subsingleton` instances for `‖α ↪ β‖`
+-- for when they either have equal cards or `α` is empty?
```
> for when they either have equal cards

No, because that's not a true statement

> or `α` is empty?

We don't have a typeclass for "empty" yet.

ericrbg replied:
my brain clearly melted when I wrote that comment, thank you

eric-wieser commented on src/logic/embedding.lean:
```diff
@@ -83,6 +83,16 @@ protected def congr {α : Sort u} {β : Sort v} {γ : Sort w} {δ : Sort x}
 (equiv.to_embedding e₁.symm).trans (f.trans e₂.to_embedding)
 
+/-- Embeddings are equivalent under equivalences. -/
+protected def equiv {α : Sort u} {β : Sort v} {γ : Sort w} {δ : Sort x}
```
I think this should probably be called `equiv.embedding_congr`, since it's essentially `equiv.arrow_congr` but for `↪` instead of `→`. You should probably add the four `_comp`, `_refl`, `_trans`, and `_symm` lemmas that `arrow_congr` has to go with it too.

ericrbg replied:
alright, will get on with that!

ericrbg replied:
okay, so that's been done in the latest commit, but I have a couple followup q's/notes:

* is it worth putting `@[congr, simps apply]` on this def too? I'm shaky on how both of them work; I put them in as a parallel to `arrow_congr`.
* `arrow_congr` uses `Sort*`, whilst `logic.embedding` tends to have universe args. I went with `Sort*`; does it matter?
* `equivalence` has no non-coe `comp` (unless I'm blind, which wouldn't surprise me) so I skipped `embedding_congr_comp`,

eric-wieser replied:
> equivalence has no non-coe comp

I assume you meant `embedding`? It does, but confusingly it's called `trans`. So perhaps that lemma would be called `embedding_congr_apply_trans` to distinguish it from `embedding_congr_trans`?

eric-wieser commented on src/logic/embedding.lean:
```diff
@@ -93,6 +103,14 @@ protected noncomputable def equiv_of_surjective {α β} (f : α ↪ β) (hf : su
+{ to_fun := λ f, ⟨f.val, f.property⟩,
+  inv_fun := λ f, ⟨f, f.injective⟩,
+  left_inv := λ f, by simp,
+  right_inv := λ f, by {ext, simp} }
```
This feels like a weird definition to have; mathlib has a lot of types that are just a single data field with some proof fields, but we don't normally bother to construct equivalences with subtypes. Perhaps it's worth just constructing this locally in the file it's needed, and making it `private`?

ericrbg replied:
dope, wasn't sure whether this would be of general use

ericrbg committed:
lintfix

ericrbg committed:
deal with most of PR stuff

ericrbg committed:
add a better `embedding_congr` API

ericrbg committed:
Apply suggestions from code review

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ericrbg committed:
Lintfix

ericrbg committed:
Fixup a little more formatting

ericrbg committed:
make filename consistent

eric-wieser commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,169 @@
+  right_inv := λ f, by {ext, simp} }
+
+-- `decidable_pred (@injective α β)` and various variations didn't give me an instance 🤷‍♂️
+noncomputable instance embedding {α β} [fintype α] [fintype β] : fintype (α ↪ β) :=
```
```suggestion
instance embedding {α β} [fintype α] [fintype β] [decidable_eq α] [decidable_eq β] :
  fintype (α ↪ β) :=
```

eric-wieser commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,169 @@
+This file establishes the cardinality of `α ↪ β` in full generality.
+-/
+
+open_locale classical
```
I think you should try to remove this - it's going to cause problems with the computable instance I propose. You can replace it with:

* The definition in #7394
* Some `[decidable_eq α] [decidable_eq β]` arguments below

ericrbg replied:
fingers crossed the last commit solved this, but can't check until CI wants to :(

eric-wieser commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,169 @@
+
+  -- type-checker doesn't like just using fin.cons
+  let extend : (fin n → β) → β → fin n.succ → β :=
+    λ f : fin n → β, λ b : β, fin.cons b f,
```
```suggestion
  let extend : (fin n → β) → β → fin n.succ → β := λ f b, fin.cons b f,
```

eric-wieser commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,169 @@
+    apply le_antisymm,
+    { by_contra card_too_big, let first := λ g : fin n.succ ↪ β, g 0,
+      -- (if I just write down x 0 = y 0, it won't work! need this shim)
+      suffices : ∃ x ∈ equiv_classes f, ∃ y ∈ equiv_classes f, x ≠ y ∧ first x = first y,
```
```suggestion
    { by_contra card_too_big,
      suffices : ∃ (x : fin n.succ ↪ β) (hx : x ∈ equiv_classes f)
                   (y : fin n.succ ↪ β) (hy : y ∈ equiv_classes f), x ≠ y ∧ x 0 = y 0,
```
and remove all references to `first` below

eric-wieser commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,169 @@
+        { rintros i - eq, simp only [extend, fin.cons_zero, fin.cons_succ] at eq,
+        have := x.prop, rw mem_poss_vals at this, exfalso, apply this i, exact eq.symm},
+        { rintros i - eq, simp only [extend, fin.cons_zero, fin.cons_succ] at eq,
+        have := x.prop, rw mem_poss_vals at this, exfalso, apply this i, exact eq},
```
Mathlib style doesn't like multiple tactics in the same line:

```suggestion
        { rintros i - eq,
          simp only [extend, fin.cons_zero, fin.cons_succ] at eq,
          have := x.prop,
          rw mem_poss_vals at this,
          exfalso,
          apply this i,
          exact eq.symm},
        { rintros i - eq,
          simp only [extend, fin.cons_zero, fin.cons_succ] at eq,
          have := x.prop,
          rw mem_poss_vals at this,
          exfalso,
          apply this i, exact eq},
```
Same throughout

ericrbg replied:
Alright, that makes sense! I liked my "logical units" approach but honestly even looking at this it feels far more readable

kim-em commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +143,37 @@ begin
+section desc_fac
+
+-- is there some way to tell only `norm_num` to expand this out, but not `simp`?
+/-- desc_fac n k = (n + k)! / n!, but implemented in a recursive way for calculation. -/
```
Could you add a sentence to the doc-string explaining the relationship with `pochhammer`? (Perhaps even a sentence at `pochhammer`, too?)

kim-em commented on archive/100-theorems-list/93_birthday_problem.lean:
```diff
@@ -0,0 +1,22 @@
+
+As of writing, probability is not implemented in Lean, so we instead state the birthday problem
+in terms of injective functions. The general result about `fintype.card (α ↪ β)` which this proof
+uses is `fintype.card_embedding`.
```
I am speaking from ignorance about this part of the library here, but we have counting measure, and of course we have measures of sets, so it is perfectly possible to state things like "the measure of the injective functions is < 1/2". It just seems an overly strong statement "probability in not implemented in Lean"!

ericrbg replied:
I wasn't actually aware of this; I read somewhere that "lean was not ready for probability" but clearly was misguided!

kim-em replied:
Certainly it has plenty of *measure theory*. Perhaps someone was trying to make the distinction between measure theory per se and how probabilists think in terms of random variables, with "no underlying measure space".

ericrbg committed:
Merge branch 'master' into birthday

ericrbg committed:
Revert "Merge branch 'master' into birthday"

This reverts commit 6de6b382618075d6770b9dc69f2af0c4878f01ca.

ericrbg committed:
cleanup of comments and other random bits and bobs

ericrbg committed:
rewrite to mathlib style and hope I can get a build from CI

ericrbg committed:
quicksave before merge

ericrbg committed:
Merge remote-tracking branch 'origin/master' into birthday

ericrbg committed:
finishup changes, make sure CI works!

eric-wieser commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,217 @@
+
+instance embedding {α β} [fintype α] [fintype β] [decidable_eq α] [decidable_eq β] :
+  fintype (α ↪ β) :=
+  fintype.of_equiv {f : α → β // function.injective f} (equiv_inj_subtype α β)
```
I'd probably put these in `data/fintype/basic`. I really hope there's a nicer way to do this though - since ideally we'd also have `fintype (α →+ β)`, `fintype (α →* β)`, `fintype (α →+* β)` instances, and making a trivial equiv to a subtype each time is tedious.

ericrbg replied:
yeah, and this approach also doesn't work for `infinite ↪ finite`, which is obviously empty. i'll have a think about this and see if I can create some nicer approach

eric-wieser replied:
I don't think you can get around needing an entirely separate instance to handle `finite ↪ finite` and `infinite ↪ finite`.

ericrbg replied:
yeah, that's what I was just going to say after messing around with it. `[decidable (infinite α)]` gets around it but that is possibly one of the most useless typeclass arguments in existence

i've also not figured any way to avoid just doing the subtype kerfuffle, i'll work on a separate PR for all the sorts of arrows later and see if they lend themselves to a nicer approach. i'll try cover the list that you covered on your PR

ericrbg replied:
okay, so I can write it as this:

`instance embedding' {α β : Type*} [fintype β] [decidable_eq β] [decidable (infinite α)] :
  fintype (α ↪ β) := if h : infinite α then sorry else sorry`

or I can just do two separate instances. my thinking for why this may be okay is that it works well with `open_locale classical`, but I think it'd also need me to create two instances which basically turn `infinite` or `fintype` into `decidable (infinite α)`. or I could just make an `infinite α` instance; how well-used is `infinite` in mathlib?

eric-wieser replied:
`[decidable (infinite α)]` sounds like a bad idea to me.

ericrbg replied:
Just realised - we could use `fintype.of_injective` for the other homs (coe to fun), but that's noncomputable. Thoughts?

ericrbg replied:
@eric-wieser, was gonna work on this today; thoughts on `fintype.of_injective`? not sure how much of a big deal noncomputablility is for `→+, →*, →+*`

eric-wieser replied:
I think a better approach would be to construct the tedious subtype equivs to make it computable. I think noncomputable fintype instances are probably ok in the classical locale, but I wouldn't want to end up using them by accident.

eric-wieser commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,217 @@
+
+/-- If `‖β‖ < ‖α‖` there is no embeddings `α ↪ β`. This is the pigeonhole principle. -/
+@[simp] theorem card_embedding_eq_zero (h : ‖β‖ < ‖α‖)
+  : ‖α ↪ β‖ = 0 :=
```
A style point: `:` always goes at the end of the previous line
```suggestion
@[simp] theorem card_embedding_eq_zero (h : ‖β‖ < ‖α‖) :
  ‖α ↪ β‖ = 0 :=
```
or all on one line:
```suggestion
@[simp] theorem card_embedding_eq_zero (h : ‖β‖ < ‖α‖) : ‖α ↪ β‖ = 0 :=
```

eric-wieser commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,217 @@
+  have := f.injective fne, contradiction
+end
+
+theorem card_embedding_eq_iff: ‖α ↪ β‖ = if ‖α‖ ≤ ‖β‖ then nat.desc_fac (‖β‖ - ‖α‖) ‖α‖ else 0 :=
```
```suggestion
theorem card_embedding_eq_if : ‖α ↪ β‖ = if ‖α‖ ≤ ‖β‖ then nat.desc_fac (‖β‖ - ‖α‖) ‖α‖ else 0 :=
```

eric-wieser commented on src/logic/embedding.lean:
```diff
@@ -232,6 +232,36 @@ end function
+@[congr, simps apply] def embedding_congr {α β γ δ : Sort*}
+  (h : α ≃ β) (h' : γ ≃ δ) : (α ↪ γ) ≃ (β ↪ δ) :=
+{ to_fun := λ f,
+    ⟨h' ∘ f ∘ h.symm, (h'.comp_injective _).mpr $ (h.symm.injective_comp _).mpr f.injective⟩,
```
Does something like this work?
```suggestion
    h.symm.to_embedding.trans (f.trans h'.to_embedding),
```

ericrbg replied:
yep, it works! also simplifies `embedding_congr_refl` to `ext, refl` :D

eric-wieser commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,217 @@
+        simp only [function.embedding.coe_fn_mk, subtype.coe_mk] at g_extended,
+        refine ⟨a, a_poss, _⟩,
+        rw ←g_extended,
+        simp,
```
```suggestion
      { intros g g_extended,
        simp only [extended, mem_map] at g_extended,
        obtain ⟨⟨a, a_poss⟩, -, g_extended⟩ := g_extended,
        simp only [function.embedding.coe_fn_mk, subtype.coe_mk] at g_extended,
        refine ⟨a, a_poss, _⟩,
        rw ←g_extended,
        simp, },
```

ericrbg committed:
all changes except moving to `fintype`

ericrbg committed:
moving the instances to `fintype`

ericrbg committed:
fixlint as I'm an idiot

ericrbg committed:
this will compile bc fk sense

ericrbg committed:
should compile now

eric-wieser commented on src/data/fintype/basic.lean:
```diff
@@ -1019,6 +1019,16 @@ fintype.of_surjective quotient.mk (λ x, quotient.induction_on x (λ x, ⟨x, rf
+
+instance function.embedding.fintype {α β} [fintype α] [fintype β]
+  [decidable_eq α] [decidable_eq β] : fintype (α ↪ β) :=
+  fintype.of_equiv _ (equiv_inj_subtype α β)
```
```suggestion
fintype.of_equiv _ (equiv_inj_subtype α β)
```

eric-wieser commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,217 @@
+    exact card_embedding_eq_zero (not_le.mp h)
+end
+
+-- it feels weird repeating the proof but I think it can only be this way due to proof irrelevance?
```
You shouldn't need to do this; all you need to prove is that `finset.empty.card = 0`

ericrbg replied:
I get that `fintype` is a subsingleton and so my instance is = to all other instances, but how'd I actually turn that into a proof? struggling to find some `rw`s that will get back to my definition

eric-wieser replied:
You don't even need it to be subsingleton, if you `set_option pp.implicit true` you'll notice that the term in this statement is literally your instance. I'll be able to help once github has a cache to give me...

ericrbg replied:
figured it out with `pp.implicit`! it's really useful with typeclasses, wow

ericrbg committed:
made `card_embedding_eq_infinite` slicker

ericrbg committed:
Update src/data/fintype/basic.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ericrbg committed:
style

ericrbg committed:
typo

tb65536 commented on src/data/fintype/basic.lean:
```diff
@@ -1500,6 +1510,11 @@ assume (hf : surjective f),
 have H : infinite α := infinite.of_surjective f hf,
 @infinite.not_fintype _ H infer_instance
 
+instance function.embedding.fintype' {α β} [infinite α] [fintype β] : fintype (α ↪ β) :=
```
Is there an "empty" instance that implies `fintype`?

ericrbg replied:
I don't know if there's empty, but `subsingleton` will certainly also do here!

tb65536 commented on src/data/fintype/basic.lean:
```diff
@@ -1019,6 +1019,16 @@ fintype.of_surjective quotient.mk (λ x, quotient.induction_on x (λ x, ⟨x, rf
+
+instance function.embedding.fintype {α β} [fintype α] [fintype β]
+  [decidable_eq α] [decidable_eq β] : fintype (α ↪ β) :=
+fintype.of_equiv _ (equiv_inj_subtype α β)
```
Could you combine the two versions of `function.embedding.fintype` into a single version without `fintype α`?

ericrbg replied:
yep, but not decidably; I had a discussion with Eric on this earlier. I will add a non-computable instance too, though, like how Scott suggested

kim-em commented on src/data/fintype/basic.lean:
```diff
@@ -1019,6 +1019,16 @@ fintype.of_surjective quotient.mk (λ x, quotient.induction_on x (λ x, ⟨x, rf
 instance finset.fintype [fintype α] : fintype (finset α) :=
 ⟨univ.powerset, λ x, finset.mem_powerset.2 (finset.subset_univ _)⟩
 
+private def equiv_inj_subtype (α β : Sort*) : {f : α → β // function.injective f} ≃ (α ↪ β) :=
```
Why is this `private`? Really things shouldn't be `private` in mathlib unless in exceptional circumstances. It's saying that you're _certain_ no one will ever want to use it.

kim-em replied:
name should probably be `subtype_injective_equiv_embedding`?

b-mehta replied:
If anything I think this could be helpful - maybe in `data/equiv/basic`?

eric-wieser replied:
I suggested this be private because I can't think of any situation you would ever need it. It's only used here as glue to the fintype api. If `private` is scary, then maybe we could inline it into the proof where it's used?

b-mehta replied:
> I suggested this be private because I can't think of any situation you would ever need it.

I disagree - I think it's helpful to have equivalences like these comparing explicit subtypes to bundled functions, especially when working with chains of equivalences. Here, it doesn't even relate to fintypes specifically: I think it should be a public lemma and be added to the API for embeddings or equivalences.

kim-em replied:
Yes, I agree this is a useful equivalence should be part of the API.

ericrbg replied:
I moved this to `logic/embedding` under the `equiv` namespace, and made it public!

kim-em commented on src/data/fintype/basic.lean:
```diff
@@ -1500,6 +1510,11 @@ assume (hf : surjective f),
 have H : infinite α := infinite.of_surjective f hf,
 @infinite.not_fintype _ H infer_instance
 
+instance function.embedding.fintype' {α β} [infinite α] [fintype β] : fintype (α ↪ β) :=
```
Can you please add

```
noncomputable def fintype_of_not_infinite {α : Type*} (h : ¬ infinite α) : fintype α :=
begin
  rw [←not_nonempty_fintype, not_not] at h,
  exact classical.choice h,
end

noncomputable
instance function.embedding.fintype'' {α β : Type*} [fintype β] : fintype (α ↪ β) :=
begin
  by_cases h : infinite α,
  { resetI, apply_instance, },
  { haveI := fintype_of_not_infinite h, classical, apply_instance, },
end
```

kim-em replied:
The proof of `fintype.exists_infinite_fiber` should be refactored to use `fintype_of_not_infinite`.

ericrbg replied:
should I put a lower instance priority on the `noncomputable` instance or will it do so automatically?

eric-wieser replied:
I've run into pain in the past with noncomputable fintype instances, and dont really think they are a good idea. For one, they prevent lean from evaluating sums over small embedding types

ericrbg replied:
is it worth possibly not making these an explicit instance, but an available defn for people to use if required?

kim-em replied:
Hmm... I'm skeptical about avoiding classical `fintype` instances in principle. It's not very mathlib-y. :-)

Unless it actually breaks something already in mathlib, I'd propose punting on any computational problems. Perhaps, @eric-wieser, you could explain your explain over on zulip?

kim-em commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,217 @@
+        apply fin.induction_on a₁;
+        apply fin.induction_on a₂,
+        -- not sure how to do this `induction_on` using the `induction` tactic
+        -- (all variants of `induction using fin.something` didn't work for me)
```
That's fine: `induction using` is pretty unreliable, and mathlib extensively just uses `apply` for this.

tb65536 commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,217 @@
+    exfalso,
+    apply ne,
+    ext x,
+    exact x.elim0 },
```
Here's a shorter proof of the base case:
```suggestion
    { nontriviality (fin 0 ↪ β),
      rw [nat.desc_fac_zero, fintype.card_eq_one_iff],
      refine ⟨nonempty.some nontrivial.to_nonempty, λ x, function.embedding.ext fin.elim0⟩ },
```

kim-em commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,217 @@
+
+namespace fintype
+
+private lemma card_embedding_aux (n : ℕ) (β) [fintype β] [decidable_eq β] (h : n ≤ ‖β‖) :
+  ‖fin n ↪ β‖ = nat.desc_fac (‖β‖ - n) n :=
```
I'm wondering if this would work more easily by the following argument:
* `X ⊕ Y ↪ Z` is equivalent to pairs of embeddings `X ↪ Z` and `Y ↪ Z` with disjoint images.
* an embedding with image disjoint from some set is equivalent to an embedding to the complement of that set (perhaps prove the function version first, or find it?)
* a calculation of the cardinality of a complement (hopefully already available)

It would certainly be more modular / reusable.

kim-em replied:
(Oh, and then to do the induction step here by replacing `fin (n+1)` with `fin n ⊕ unit`.)

ericrbg replied:
Scott, I tried some of this in [here](https://github.com/leanprover-community/mathlib/blob/2a87ca3b791da24c05b4dfe20d8be07410b9ae51/src/data/fintype/card_embedding.lean); would appreciate some comments on whether `function_thing` and `to_be_named` should be split up more, or everything should be defined inline.

as a bonus, though, this refactor seems to make the proof a looot shorter

kim-em replied:
`function_thing` already exists as `sum_arrow_equiv_prod_arrow` (I found this by typing `⊕.*→.*≃` in the search box in VSCode). `to_be_named` looks good.

kim-em replied:
I suspect that splitting `to_be_named` according to the first two bullet points I suggested above will be even easier (as then you can use `function_thing` for the "data" part of the equivalence. But perhaps not; try both ways?

ericrbg replied:
I tried a little splitting that up, but I wasn't sure how to express the disjointness too cleanly except as a subtype with disjoint set.ranges. I'll give it another try though!

eric-wieser replied:
A more symmetric formulation of `to_be_named` might be `((α ⊕ β) ↪γ) ≃ (Σ s : set γ, α ↪ s × β↪ s.compl)`

ericrbg replied:
I'll put that on as a corollary, as I think it's probably of more general use, but it won't work in this case as it doesn't give me the size of `s`

ericrbg replied:
Alright, the new version is now up. It's not done yet (naming, docstrings, Eric's corollary) but the idea behind it is now clear; I have some questions in the new comments

kim-em replied:
> I tried a little splitting that up, but I wasn't sure how to express the disjointness too cleanly except as a subtype with disjoint set.ranges. I'll give it another try though!

Did this not work? The `disjoint (set.range f) (set.range g)` version seems nicest to prove first. But I haven't actually tried it.

ericrbg replied:
just did it now scott; was scared of it and ended up being alright!

ericrbg committed:
some ideas, not sure how well they'll work

ericrbg committed:
overview of idea

ericrbg committed:
add more sorries

ericrbg committed:
finishup to_be_named

ericrbg committed:
better version

ericrbg committed:
sorry-free!

kim-em commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,156 @@
+  rw card_congr to_be_named,
+  all_goals { try { apply_instance } },
+
+  have : ∀ (f : fin n ↪ β), ‖fin 1 ↪ ↥((set.range f).compl)‖ = ‖β‖ - n,
```
This should certainly be a separate lemma, just computing `‖fin 1 ↪ X‖ = card X`.

eric-wieser replied:
Or more generally, `card (T ↪ X) = card X` given `[unique T]`.

ericrbg replied:
didn't know about `[unique]` - thanks!

ericrbg replied:
okay, this lemma is cursed:

```lean
lemma card_embedding_of_unique {α β : Type*} [unique α] [fintype β] [decidable_eq β]:
‖α ↪ β‖ = ‖β‖ := card_congr some_other_nameless_thing
```

If I even dare put it in the simpset, it causes a timeout (even if I comment out all simps apart from `nontriviality`! So it shouldn't even be looking there!). Trying to `rw` with it also causes Lean to give up, but even more weirdly, `card_congr some_other_nameless_thing` is just fine. Maybe it's the subsingleton instances again, but I'm really tempted not to put it in as it's toxic waste

eric-wieser replied:
Maybe try adding `[fintype α]`?

ericrbg replied:
doesn't help :(

eric-wieser replied:
What fintype instance is the statement using? Can you push the failing definition to a branch so I can try it out in context?

Another idea might be to use `[subsingleton α] [nonempty α] [fintype α]` instead.

ericrbg replied:
https://github.com/leanprover-community/mathlib/commit/3ce5e10a706e23d90feb32f505a84802307c39e0 commit with issues here

kim-em replied:
What's the status here? Maybe you can start a topic on zulip to ask for help if necessary?

ericrbg replied:
I'll try bring it up to date, play around with it a little, and then ask in Zulip if not

ericrbg replied:
okay, with the new code, it _seems_ to work just fine locally! hopefully CI agrees

kim-em commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,156 @@
+    rw card_of_finset' (finset.map f univ)ᶜ,
+    { simp [card_compl, card_map] }, -- this line does all our cardinality stuff for us!
+    -- next, we prove that `x ∈ map f univ ↔ x ∈ set.range f`. should probably be extracted, but:
+    -- not sure to where, and not sure what to call it. also, feel like I must've missed it
```
This part certainly shouldn't be necessary. Ask for help if you can't find it.

ericrbg replied:
argh `simp` can do it with the `change`, but still not sure why I can't use the `\compl` in the statement

ericrbg replied:
@semorrison this is on an outdated file so it's not on the file list, but I'm still not sure why `\compl` doesn't work and I have to mess around with both `.compl` and `\compl`; any ideas?

ericrbg committed:
most changes, todo: add subtype version

matt-kempster commented on docs/100.yaml:
```diff
@@ -343,6 +343,10 @@
   title  : Pick’s Theorem
 93:
   title  : The Birthday Problem
+  decl   : fintype.card_embedding
```
This should probably be removed now that there is a `mathlib archive` link, right?

ericrbg replied:
I wanted to keep it this way, so that people could see where it actually comes from. The proof in `archive` is opaque almost by design :)

ericrbg committed:
sorry free, stripped `opens`

ericrbg committed:
version before moving everything. should build!

ericrbg committed:
CI-driven development :)

ericrbg committed:
first error bites the dust

ericrbg committed:
added an import to make things work nicer

ericrbg committed:
move some advanced equivs to a separate file

ericrbg committed:
fixup and import new file

kim-em commented on src/data/equiv/embedding.lean:
```diff
@@ -0,0 +1,84 @@
+
+/-- Embeddings from a sumtype are equivalent to two separate embeddings with disjoint ranges. -/
+def sum_embedding_equiv_prod_embedding_disjoint {α β γ : Type*} :
+((α ⊕ β) ↪ γ) ≃ {f : (α ↪ γ) × (β ↪ γ) // disjoint (set.range f.1) (set.range f.2)} :=
```
```suggestion
  ((α ⊕ β) ↪ γ) ≃ {f : (α ↪ γ) × (β ↪ γ) // disjoint (set.range f.1) (set.range f.2)} :=
```

kim-em commented on src/data/equiv/embedding.lean:
```diff
@@ -0,0 +1,84 @@
+    simp only at this,
+    assumption
+  end⟩,
+
```
We don't usually leave blank lines between fields.

kim-em replied:
Still needs fixing.

ericrbg replied:
Oops, yep, on it

kim-em commented on src/data/equiv/embedding.lean:
```diff
@@ -0,0 +1,84 @@
+
+namespace equiv
+
+/-- Embeddings from a sumtype are equivalent to two separate embeddings with disjoint ranges. -/
```
```suggestion
/-- Embeddings from a sum type are equivalent to two separate embeddings with disjoint ranges. -/
```
(Going off 17 vs 0 appearances in mathlib.)

kim-em commented on src/data/equiv/embedding.lean:
```diff
@@ -0,0 +1,84 @@
+  {f : (α ↪ γ) × (β ↪ γ) // disjoint (set.range f.1) (set.range f.2)} ≃
+  (Σ f : α ↪ γ, β ↪ (set.range f).compl) :=
+{ to_fun := λ ⟨⟨f, g⟩, disj⟩, ⟨f, ⟨λ b, ⟨g b, set.disjoint_right.mp disj (⟨b, rfl⟩)⟩,
+  λ _ _ eq, by { rw subtype.mk_eq_mk at eq, exact g.injective eq }⟩⟩, -- one day I'll figure out ▸
```
```suggestion
  λ _ _ eq, by { rw subtype.mk_eq_mk at eq, exact g.injective eq }⟩⟩,
```
Don't bother, `rw` is clearer to read anyway?

kim-em commented on src/data/equiv/embedding.lean:
```diff
@@ -0,0 +1,84 @@
+  right_inv := λ ⟨_,_⟩, by simp!,
+  left_inv := λ ⟨⟨_,_⟩,_⟩, by { dsimp only, ext; simp! } }
+
+/-- A simple combination of the above results, allowing us to simply combine embeddings
```
```suggestion
/-- A combination of the above results, allowing us to combine embeddings
```
Doc-string could be more informative. What is "combine embeddings"?

kim-em commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,90 @@
+  have := f.injective fne, contradiction
+end
+
+theorem card_embedding_eq_if: ‖α ↪ β‖ = if ‖α‖ ≤ ‖β‖ then nat.desc_fac (‖β‖ - ‖α‖) ‖α‖ else 0 :=
```
```suggestion
theorem card_embedding_eq_if : ‖α ↪ β‖ = if ‖α‖ ≤ ‖β‖ then nat.desc_fac (‖β‖ - ‖α‖) ‖α‖ else 0 :=
```

kim-em commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +143,38 @@ begin
 
+section desc_fac
+
+-- is there some way to tell only `norm_num` to expand this out, but not `simp`?
```
```suggestion
```
Not for now.

ericrbg commented on src/data/equiv/embedding.lean:
```diff
@@ -0,0 +1,84 @@
+  left_inv := λ ⟨⟨_,_⟩,_⟩, by { dsimp only, ext; simp! } }
+
+/-- A simple combination of the above results, allowing us to simply combine embeddings
+over a sum-type. -/
```
```suggestion
/-- A combination of the above results, allowing us to turn one embedding over a sum type
into two dependent embeddings, the second of which avoids any members of the range
of the first. This is helpful for constructing larger embeddings out of smaller ones. -/
```

ericrbg commented on src/data/equiv/embedding.lean:
```diff
@@ -0,0 +1,84 @@
+{ to_fun := λ ⟨⟨f, g⟩, disj⟩, ⟨f, ⟨λ b, ⟨g b, set.disjoint_right.mp disj (⟨b, rfl⟩)⟩,
+  λ _ _ eq, by { rw subtype.mk_eq_mk at eq, exact g.injective eq }⟩⟩, -- one day I'll figure out ▸
+
+  inv_fun := λ ⟨f, g⟩, ⟨⟨f, g.trans (function.embedding.subtype _)⟩,
```
```suggestion
{ to_fun := λ ⟨⟨f, g⟩, disj⟩, ⟨f, ⟨λ b, ⟨g b, set.disjoint_right.mp disj (⟨b, rfl⟩)⟩,
  λ _ _ eq, by { rw subtype.mk_eq_mk at eq, exact g.injective eq }⟩⟩,
  inv_fun := λ ⟨f, g⟩, ⟨⟨f, g.trans (function.embedding.subtype _)⟩,
```

ericrbg commented on src/data/equiv/embedding.lean:
```diff
@@ -0,0 +1,84 @@
+    exact set.disjoint_of_subset_right this disjoint_compl_right
+  end⟩,
+
+  right_inv := λ ⟨_,_⟩, by simp!,
```
```suggestion
  end⟩,
  right_inv := λ ⟨_,_⟩, by simp!,
```

ericrbg committed:
Apply suggestions from code review

Co-authored-by: Scott Morrison <scott.morrison@gmail.com>

ericrbg committed:
Merge branch 'master' into birthday

ericrbg committed:
Move desc_fac out

ericrbg committed:
fix changes to equiv_fin

ericrbg committed:
golf

Co-authored-by: Benjamin Davidson <68528197+benjamindavidson@users.noreply.github.com>

ericrbg committed:
golf right

ericrbg committed:
Merge branch 'master' into desc_fac

ericrbg committed:
Please let me have both CI runs!

ericrbg committed:
try adding @[simp]

ericrbg committed:
Merge branch 'desc_fac' into birthday

ericrbg committed:
not make an absolute disaster out of everything

ericrbg committed:
this is a nightmare; think this is the last one

ericrbg committed:
Try removing @[simp]

ericrbg committed:
should finally be good

ericrbg committed:
it should not be a `simp` lemma

ericrbg committed:
improve proof

ericrbg committed:
`ac_refl` begone

Co-authored-by: Benjamin Davidson <68528197+benjamindavidson@users.noreply.github.com>

ericrbg committed:
docstring updates

ericrbg committed:
Update src/data/nat/factorial.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

ericrbg committed:
avoid 100-char linter

ericrbg committed:
update docs/names

ericrbg committed:
Apply suggestions from code review

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ericrbg committed:
turns out I use lemmas I write

ericrbg committed:
:(

ericrbg committed:
use `rw_mod_cast`

ericrbg committed:
more rw_mod_cast and stop `rw`ing away `succ`

I'm in a bad habit with that because of `ring` not dealing with `succ` well...

Co-authored-by: Benjamin Davidson <68528197+benjamindavidson@users.noreply.github.com>

ericrbg committed:
improve readability

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ericrbg committed:
split `eval_succ`

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ericrbg committed:
Merge branch 'desc_fac' into birthday

ericrbg committed:
update to newer `desc_fac` api

ericrbg committed:
Merge branch 'master' into birthday

ericrbg committed:
remove spaces in structure

ericrbg committed:
hooray! things seem to be fine now

ericrbg committed:
Trigger Build

kim-em commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,94 @@
+
+-- the [fintype α] argument may seem redundant; it isn't!
+lemma card_embedding_of_unique
+{α β : Type*} [unique α] [fintype α] [fintype β] [decidable_eq α] [decidable_eq β]:
```
```suggestion
  {α β : Type*} [unique α] [fintype α] [fintype β] [decidable_eq α] [decidable_eq β] :
```

kim-em commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,94 @@
+
+namespace fintype
+
+-- the [fintype α] argument may seem redundant; it isn't!
```
Wait, why?

eric-wieser replied:
Because without it, `unique.fintype` is inferred in the lemma statement and it doesn't match against situations where another `fintype` instance is used.

kim-em replied:
Ok. Could you just replace the comment with something that says this? e.g.
```
-- We need the separate `fintype α` instance as it contains data,
-- and may not match definitionally with the instance coming from `unique.fintype`.
```

kim-em commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,94 @@
+  have : ∀ (f : fin n ↪ β), ‖fin 1 ↪ ↥((set.range f).compl)‖ = ‖β‖ - n,
+  {
+    intro f,
+    rw card_embedding_of_unique; try {apply_instance},
```
I think this is abuse of `;` and `try`. Just use `apply_instance` where it is needed.

ericrbg replied:
ahh, I can actually remove this now; originally `card_embedding_of_unique` (or the `card_congr` lemma) just refused to use inference, and actually instead set them as subgoals; now that it's packaged up, inference ends up working fine

kim-em commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,94 @@
+
+  have : ∀ (f : fin n ↪ β), ‖fin 1 ↪ ↥((set.range f).compl)‖ = ‖β‖ - n,
+  {
+    intro f,
```
Per style guide, braces don't deserve their own line.

kim-em commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,94 @@
+    rw card_of_finset' (finset.map f finset.univ)ᶜ,
+    { simp [finset.card_compl, finset.card_map] },
+    { -- further evidence `ᶜ` is defeq, something odd
+    change ∀ x, x ∈ (finset.map f finset.univ)ᶜ ↔ x ∈ (set.range ⇑f)ᶜ,
```
Are we missing a `simp` lemma that rewrites `set.compl` in terms of the general boolean algebra complement?

ericrbg replied:
I managed to get around this (turns out if I _explicitly_ use the `coe_to_sort` arrow I can write `\compl` in the definitions) but I'm still gonna post in Zulip about this because `.compl` has given me many headaches.

kim-em commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,94 @@
+    intro f,
+    rw card_embedding_of_unique; try {apply_instance},
+    rw card_of_finset' (finset.map f finset.univ)ᶜ,
+    { simp [finset.card_compl, finset.card_map] },
```
Either change this to
```
    { simp only [finset.card_compl, finset.card_map, finset.card_fin], },
```
or, perhaps, better, investigate if mathlib is happy with `finset.card_compl` becoming a simp lemma, and just using `{ simp }`.

ericrbg replied:
I think you'd sometimes often work with `\compl` just to rewrite it to something else, so I don't think it's too sensible of a simp-lemma. I just changed this to an explicit `rw`

kim-em commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,94 @@
+    exact h,
+    exact nat.succ_pos _ },
+
+  rw [←ht, hn (nat.lt_of_succ_le h).le, h', mul_comm, nat.succ_desc_fac, nat.desc_fac_succ]
```
Could we restructure this so the block starts with
```
rw hn (nat.le_of_succ_le h),
```
making it clear that we're done with the actual induction, and just have some final mucking about with `desc_fac`?

Ideally actually you could pull out some or all of that final cleanup to a lemma in `data.nat.factorial`.

kim-em commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,94 @@
+  exact card_embedding_aux _ _ h,
+end
+
+/-- If `‖β‖ < ‖α‖` there is no embeddings `α ↪ β`. This is the pigeonhole principle. -/
```
```suggestion
/-- If `‖β‖ < ‖α‖` there are no embeddings `α ↪ β`. This is a formulation of the pigeonhole principle. -/
```

kim-em commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,94 @@
+end
+
+lemma card_embedding_eq_infinite {α β} [infinite α] [fintype β] : fintype.card (α ↪ β) = 0 :=
+by erw [fintype.card, finset.univ, function.embedding.fintype', finset.card_empty]
```
I am completely mystified by this proof! (That's not the end of the world, but how is this working?)

eric-wieser replied:
I think `exact finset.card_empty` might be enough to close this goal

kim-em replied:
Oh!

ericrbg replied:
`finset.card_empty` is `rfl` so this is actually a `rfl` lemma! I'll leave a note explaining that; this uses the `function.embedding.fintype'` instance, which is definitionally empty

kim-em commented on src/logic/embedding.lean:
```diff
@@ -232,6 +235,45 @@ end function
 
+open function.embedding
+
+/-- The type of injections α ↪ β is equivalent to the subtype of all injective functions α → β. -/
```
```suggestion
/-- The type of embeddings `α ↪ β` is equivalent to the subtype of all injective functions `α → β`. -/
```

ericrbg committed:
tidy up from code review

ericrbg committed:
fixup `\compl` and `desc_fac`

ericrbg committed:
improve docs

Co-authored-by: Scott Morrison <scott.morrison@gmail.com>

ericrbg committed:
lintfix

eric-wieser commented on src/data/fintype/card_embedding.lean:
```diff
@@ -66,7 +66,8 @@ begin
 
-/-- If `‖β‖ < ‖α‖` there are no embeddings `α ↪ β`. This is a formulation of the pigeonhole principle. -/
+/-- If `‖β‖ < ‖α‖` there are no embeddings `α ↪ β`.
+    This is a formulation of the pigeonhole principle. -/
```
```suggestion
This is a formulation of the pigeonhole principle. -/
```
I don't think docgen wants this indent.

ericrbg committed:
make docgen happy

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on src/data/equiv/embedding.lean:
```diff
@@ -0,0 +1,82 @@
+    exact set.disjoint_of_subset_right this disjoint_compl_right
+  end⟩,
+  right_inv := λ ⟨_,_⟩, by simp!,
+  left_inv := λ ⟨⟨_,_⟩,_⟩, by { dsimp only, ext; simp! } }
```
I was able to split this into chain of operations which are maybe worth having in mathlib somewhere:
```suggestion

/-- A subtype of a `prod` is equivalent to a sigma type whose fibers are subtypes. -/
def _root_.subtype_prod_equiv_sigma_subtype {α β : Type*} (p : α → β → Prop) :
  {x : α × β // p x.1 x.2} ≃ Σ a, {b : β // p a b} :=
{ to_fun := λ x, ⟨(x : α × β).1, (x : α × β).2, x.prop⟩,
  inv_fun := λ x, ⟨⟨x.1, x.2⟩, x.2.prop⟩,
  left_inv := λ x, by ext; refl,
  right_inv := λ ⟨a, b, pab⟩, rfl }

/-- Embeddings whose range lies within a set are equivalent to embeddings to that set.

This is `function.embedding.cod_restrict` as an equiv. -/
def cod_restrict (α : Type*) {β : Type*} (bs : set β) :
  {f : α ↪ β // ∀ a, f a ∈ bs} ≃ (α ↪ bs) :=
{ to_fun := λ f, (f : α ↪ β).cod_restrict bs f.prop,
  inv_fun := λ f, ⟨f.trans (function.embedding.subtype _), λ a, (f a).prop⟩,
  left_inv := λ x, by { ext, refl },
  right_inv := λ x, by { ext, refl } }

/-- Pairs of embeddings with disjoint ranges are equivalent to a dependent sum of embeddings,
in which the second embedding cannot take values in the range of the first. -/
def prod_embedding_disjoint_equiv_sigma_embedding_restricted {α β γ : Type*} :
  {f : (α ↪ γ) × (β ↪ γ) // disjoint (set.range f.1) (set.range f.2)} ≃
  (Σ f : α ↪ γ, β ↪ ↥((set.range f)ᶜ)) :=
(subtype_prod_equiv_sigma_subtype $
  λ (a : α ↪ γ) (b : β ↪ _), disjoint (set.range a) (set.range b)).trans $
  equiv.sigma_congr_right $ λ a,
    (subtype_equiv_prop begin
      ext f,
      rw [←set.range_subset_iff, set.subset_compl_iff_disjoint],
      exact disjoint.comm.trans disjoint_iff,
    end).trans (cod_restrict _ _)
```

ericrbg replied:
this is really amazing! maybe the top one can go in `data.equiv.basic`, the second one I'm torn on whether it should stay here or be moved to be near `function.embedding.cod_restrict`, and the third one should (probably?) stay here

ericrbg replied:
feel free to push whatever you think is best by the way - I'm really busy today but would be nice to get this though the cogs of CI

ericrbg committed:
optimize imports

ericrbg committed:
refactor `data.equiv.embedding`

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,88 @@
+begin
+  split_ifs with h,
+    exact card_embedding h,
+    exact card_embedding_eq_zero (not_le.mp h)
```
```suggestion
  { exact card_embedding h },
  { exact card_embedding_eq_zero (not_le.mp h) }
```

eric-wieser commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,88 @@
+  intro f,
+  obtain ⟨x, y, eq, fne⟩ := fintype.exists_ne_map_eq_of_card_lt f h,
+  have := f.injective fne, contradiction
+end
```
Now that we have `is_empty`, can you add the statement:
```lean
/-- If `‖β‖ < ‖α‖` there are no embeddings `α ↪ β`.
This is a formulation of the pigeonhole principle.

Note this cannot be an instance as it needs `h`. -/
@[simp] theorem _root_.function.embedding.is_empty_of_card_lt (h : ‖β‖ < ‖α‖) : is_empty (α ↪ β) :=
⟨λ f begin
  obtain ⟨x, y, eq, fne⟩ := fintype.exists_ne_map_eq_of_card_lt f h,
  have := f.injective fne, contradiction
end⟩
```

ericrbg committed:
style

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ericrbg committed:
Merge branch 'master' into birthday

ericrbg committed:
first nonempty fixup

ericrbg committed:
clarify

eric-wieser commented on src/data/fintype/card_embedding.lean:
```diff
@@ -0,0 +1,85 @@
+
+Note this cannot be an instance as it needs `h`. -/
+@[simp] theorem _root_.function.embedding.is_empty_of_card_lt (h : ‖β‖ < ‖α‖) : is_empty (α ↪ β) :=
+⟨λ f, let ⟨x, y, ne, feq⟩ := fintype.exists_ne_map_eq_of_card_lt f h in ne $ f.injective feq⟩
```
This probably belongs in another file; perhaps next to where the fintype instance on `α ↪ β` is defined?

Also, to be clear: I think your previous lemma that said `card_embedding_eq_zero : ‖α ↪ β‖ = 0` is worth having (in this file) _in addition_ to this new lemma I suggested.

ericrbg replied:
Isn't that (and I guess also the `fintype` instance) just implied by this? Or is it worth the lack of fiddling with `haveI` to have it be separate?

eric-wieser replied:
I don't think the fintype instance is relevant here.

I think `card_embedding_eq_zero` is worth having to avoid the fiddlign with `haveI`, yes

ericrbg replied:
this reminds me, actually; do you think the `|` and `‖` notation should be localised? I find it a lot easier to read than `.card`, especially considering how it makes it obvious whether we're dealing with a `fintype` or a `finset`, so I was gonna do a PR with it afterwards on the main files - does this seem reasonable?

eric-wieser commented on src/logic/embedding.lean:
```diff
@@ -62,6 +62,9 @@ lemma ext_iff {α β} {f g : embedding α β} : (∀ x, f x = g x) ↔ f = g :=
   (@mk _ _ f i : α → β) = f := rfl
 
+@[simp] lemma mk_coe {α β : Type*} (f : α ↪ β) : ∀ inj, (⟨f, inj⟩ : α ↪ β) = f :=
+λ _, by ext; simp
```
A little tidier:
```suggestion
@[simp] lemma mk_coe {α β : Type*} (f : α ↪ β) (inj) : (⟨f, inj⟩ : α ↪ β) = f :=
by { ext, simp }
```

ericrbg committed:
`is_empty` support

ericrbg committed:
don't use ∀ when I don't have to

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ericrbg committed:
`variables` + copypaste

eric-wieser commented on src/data/fintype/basic.lean:
```diff
@@ -1429,6 +1441,10 @@ let ⟨y, hy⟩ := exists_not_mem_finset ({x} : finset α) in
 
+/-- A non-infinite type is a fintype. -/
+noncomputable def fintype_of_not_infinite {α : Type*} (h : ¬ infinite α) : fintype α :=
+by rw [←not_nonempty_fintype, not_not] at h; exact h.some
```
I think this probably belongs up on line 1414 below `is_empty_fintype`

It golfs to `((not_iff_comm.mp not_nonempty_fintype).mp h).some` if you want that.

ericrbg committed:
move `fintype_of_not_infinite`, fixup

eric-wieser commented on src/data/fintype/basic.lean:
```diff
@@ -1517,6 +1544,11 @@ assume (hf : surjective f),
+instance function.embedding.fintype' {α β} [infinite α] [fintype β] : fintype (α ↪ β) :=
+{ elems := finset.empty,
+  complete := λ f, let ⟨_, _, ne, f_eq⟩ := fintype.exists_ne_map_eq_of_infinite f in
+              false.elim $ ne $ f.injective f_eq }
```
```suggestion
{ elems := finset.empty, complete := is_empty_elim }
```

eric-wieser reviewed (approved):
bors d+

ericrbg commented on src/data/fintype/basic.lean:
```diff
@@ -1517,6 +1544,11 @@ assume (hf : surjective f),
 have H : infinite α := infinite.of_surjective f hf,
 by exactI not_fintype α
 
+instance function.embedding.fintype' {α β} [infinite α] [fintype β] : fintype (α ↪ β) :=
```
```suggestion
-- the instance generated by `is_empty → subsingleton → fintype is non-computable
instance function.embedding.fintype' {α β} [infinite α] [fintype β] : fintype (α ↪ β) :=
```

eric-wieser replied:
I think probably we should add an `is_empty.fintype` instance to handle this problem elsewhere; but not need to do so in this PR.

ericrbg replied:
yeah, was going to do it when it gets merged to master

ericrbg committed:
Apply suggestions from code review

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>
